internal/querybuilder/templates.go

```diff
@@ -0,0 +1,440 @@
+func (q *__{{ .ModelName }}SQLQueryBuilder) Limit(l int) {{ .ModelName }}SQLQueryBuilder {
+	q.mode = "select"
+	q.limit = l	
+	return q
+}
+func (q *__{{ .ModelName }}SQLQueryBuilder) Offset(l int) {{ .ModelName }}SQLQueryBuilder {
+	q.mode = "select"
+	q.offset = l
+	return q
+}
+`
+
+const toRows = `
+func (m {{ .ModelName }}) Values() []interface{} {
+    var values []interface{}
+	{{ range .Fields }}
+	values = append(values, &m.{{ .Name }})
+	{{ end }}
+    return values
+}
+`
+
+const fromRows = `
+func {{ .ModelName }}sFromRows(rows *sql.Rows) ([]{{.ModelName}}, error) {
+    var {{.ModelName}}s []{{.ModelName}}
+    for rows.Next() {
+        var m {{ .ModelName }}
+        err := rows.Scan(
+            {{ range .Fields }}
+            &m.{{ .Name }},
+            {{ end }}
+        )
+        if err != nil {
+            return nil, err
+        }
+        {{.ModelName}}s = append({{.ModelName}}s, m)
+    }
+    return {{.ModelName}}s, nil
+}
+
+func {{ .ModelName }}FromRow(row *sql.Row) ({{.ModelName}}, error) {
+    if row.Err() != nil {
+        return {{.ModelName}}{}, row.Err()
+    }
+    var m {{ .ModelName }}
+    err := row.Scan(
+        {{ range .Fields }}
+        &m.{{ .Name }},
+        {{ end }}
+    )
+    if err != nil {
+        return {{.ModelName}}{}, err
+    }
+    return m, nil
+}
+`
+
+const finishers = `
+func (q *__{{ .ModelName }}SQLQueryBuilder) Update(db *sql.DB) (sql.Result, error) {
+	q.mode = "update"
+	args := append(q.setArgs, q.whereArgs...)
+	query, err := q.SQL()
+	if err != nil {
+		return nil, err
+	}
+	return db.Exec(query, args...)
+}
+
+func (q *__{{ .ModelName }}SQLQueryBuilder) Delete(db *sql.DB) (sql.Result, error) {
+	q.mode = "delete"
+	query, err := q.SQL()
+	if err != nil {
+		return nil, err
+	}
+	return db.Exec(query, q.whereArgs...)
+}
+
+func (q *__{{ .ModelName }}SQLQueryBuilder) Fetch(db *sql.DB) ([]{{ .ModelName }}, error) {
+	q.mode = "select"
+	query, err := q.SQL()
+	if err != nil {
+		return nil, err
+	}
+	rows, err := db.Query(query, q.whereArgs...)
+	if err != nil {
+		return nil, err
+	}
+	return {{.ModelName}}sFromRows(rows)
+}
+
+func (q *__{{ .ModelName }}SQLQueryBuilder) First(db *sql.DB) ({{ .ModelName }}, error) {
+	q.mode = "select"
+	q.orderby = []string{"ORDER BY id ASC"}
+	q.Limit(1)
+	query, err := q.SQL()
+	if err != nil {
+		return {{ .ModelName }}{}, err
+	}
+	row := db.QueryRow(query, q.whereArgs...)
+	if row.Err() != nil {
+		return {{ .ModelName }} {}, row.Err()
+	}
+	return UserFromRow(row)
+}
+
+
+func (q *__{{ .ModelName }}SQLQueryBuilder) Last(db *sql.DB) ({{ .ModelName }}, error) {
+	q.mode = "select"
+	q.orderby = []string{"ORDER BY id DESC"}
+	q.Limit(1)
+	query, err := q.SQL()
+	if err != nil {
+		return {{ .ModelName }}{}, err
+	}
+	row := db.QueryRow(query, q.whereArgs...)
+	if row.Err() != nil {
+		return {{ .ModelName}} {}, row.Err()
+	}
+	return UserFromRow(row)
+}
+`
+
+const queryBuilder = `
+type __{{ .ModelName }}SQLQueryBuilder struct {
```

Comment:
Why does the name of the struct start with a double underscore?

Reply:
it's to make sure user won't construct this struct on his/her own, maybe we can find beter ways of notifying user that 'you should always construct query builder using constructor function'